### PR TITLE
many: automatically restart all-snap devices after os/kernel updates

### PR DESCRIPTION
**Pull request by mvo5**
This adds the automatic reboot back to all-snap devices. Needs more tests but I want confirmation about the direction.

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -221,11 +225,24 @@ func (d *Daemon) addRoutes() {
 	d.router.NotFoundHandler = NotFound("not found")
 }
 
+var shutdownMsg = i18n.G("snapd triggered a reboot to boot into an up to date system -- temprorarily disable the reboot by running 'sudo shutdown -c'")
```
Comment (niemeyer) [inline daemon/daemon.go]: Suggested phrasing:

`"reboot scheduled to update the system - temporarily disable with 'sudo shutdown -c'"`

File: daemon/daemon.go [hunk 2]
```diff
@@ -221,11 +225,24 @@ func (d *Daemon) addRoutes() {
...
+		case state.RestartDaemon:
+			d.tomb.Kill(nil)
+		case state.RestartSystem:
+			cmd := exec.Command("shutdown", "+10", "-r", shutdownMsg)
+			if out, err := cmd.CombinedOutput(); err != nil {
+				logger.Noticef("%s", osutil.OutputErr(out, err))
+			}
+		default:
+			// FIMXE: log error instead?
+			logger.Panicf("called with unknown restartType: %v", t)
```
Comment (niemeyer) [inline daemon/daemon.go]:
```
logger.Noticef("internal error: restart handler called with unknown restart type")
d.tomb.Kill(nil)
```

File: overlord/snapstate/snapmgr.go [hunk 3]
```diff
@@ -923,10 +923,16 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
...
-	if newInfo.Type == snap.TypeOS && release.OnClassic {
+	if release.OnClassic && newInfo.Type == snap.TypeOS {
 		t.Logf("Restarting snapd...")
 		st.Unlock()
-		st.RequestRestart()
+		st.RequestRestart(state.RestartDaemon)
+		st.Lock()
+	}
+	if !release.OnClassic && (newInfo.Type == snap.TypeOS || newInfo.Type == snap.TypeKernel) {
+		t.Logf("Restarting system...")
```
Comment (niemeyer) [inline overlord/snapstate/snapmgr.go]: Both of these messages feel a bit off.. should probably say something like:

`"Requested system restart."`
`"Requested daemon restart."`

As this is just a request, which will be respected in due time, and could even be canceled.

File: overlord/state/state.go [hunk 4]
```diff
@@ -65,6 +65,14 @@ func (data customData) set(key string, value interface{}) {
 	data[key] = &entryJSON
 }
 
+type RestartType int
+
+const (
+	RestartUnset RestartType = iota
```
Comment (niemeyer) [inline overlord/state/state.go]: `RestartNone`?

File: tests/main/ubuntu-core-upgrade-core/task.yaml [hunk 5]
```diff
@@ -26,6 +26,19 @@ execute: |
         grub-editenv list | grep "snap_try_core=ubuntu-core_$(cat nextBoot).snap"
         grub-editenv list | grep 'snap_mode=try'
 
+        echo "Ensure the device is scheduled for auto-reboot"
```
Comment (niemeyer) [inline tests/main/ubuntu-core-upgrade-core/task.yaml]: Interesting.. wonder if we should have a WAIT_REBOOT spread primitive, so we don't actually reboot but rather wait for it to be rebooted.

**Comment by niemeyer** [comment]
Just trivials, and LGTM. Very nice.

**Comment by pedronis** [comment]
msgs aside, lgtm

**Pull request merged**